cmd/ecr-credential-provider/main.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+
+	plugin "k8s.io/cloud-provider-aws/cmd/ecr-credential-provider/framework"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubelet/pkg/apis/credentialprovider/v1alpha1"
+)
+
+var ecrPattern = regexp.MustCompile(`^(\d{12})\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.(amazonaws\.com(\.cn)?|sc2s\.sgov\.gov|c2s\.ic\.gov)$`)
+
+// ECR abstracts the calls we make to aws-sdk for testing purposes
+type ECR interface {
+	GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error)
+}
+
+// ECRProvider is responsible for instantiating a new ECR instance
+type ECRProvider interface {
+	NewECR(region string, registryID string) (*ecr.ECR, error)
```

Comment:
This probably doesn't need to be an interface unless there's a need to mock it in the unit tests.

Reply:
(in general, constructor functions return implementations of an interface, not part of an interface itself)

Reply:
This is indeed for unit tests specifically.

Reply:
Where is `NewECR` being invoked? I didn't see it anywhere, maybe I missed it

Reply:
I'm using it [here](https://github.com/kubernetes/cloud-provider-aws/pull/157/files#diff-9c626dd1b5e98f480c58e2aa13e2c92d4b1cc4eee0eaa08800826ad67a29afb8R66). 

I can't make it a struct, but it works if I delete the `NewECR` method. Hmm. Let me see if I can get rid of it.

Reply:
I also noticed that the `NewMockECRProvider` sets the `ecrProvider` field but I don't see `ecrProvider` being used.